leshan-core/src/main/java/org/eclipse/leshan/core/node/codec/senml/LwM2mNodeSenMLJsonDecoder.java

```diff
@@ -0,0 +1,327 @@
+/*******************************************************************************
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *     Boya Zhang - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.leshan.core.node.codec.senml;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.eclipse.leshan.core.model.LwM2mModel;
+import org.eclipse.leshan.core.model.ResourceModel;
+import org.eclipse.leshan.core.model.ResourceModel.Type;
+import org.eclipse.leshan.core.node.LwM2mMultipleResource;
+import org.eclipse.leshan.core.node.LwM2mNode;
+import org.eclipse.leshan.core.node.LwM2mObject;
+import org.eclipse.leshan.core.node.LwM2mObjectInstance;
+import org.eclipse.leshan.core.node.LwM2mPath;
+import org.eclipse.leshan.core.node.LwM2mResource;
+import org.eclipse.leshan.core.node.LwM2mSingleResource;
+import org.eclipse.leshan.core.node.codec.CodecException;
+import org.eclipse.leshan.senml.SenMLJson;
+import org.eclipse.leshan.senml.SenMLPack;
+import org.eclipse.leshan.senml.SenMLRecord;
+import org.eclipse.leshan.util.Base64;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LwM2mNodeSenMLJsonDecoder {
+
+    private static final Logger LOG = LoggerFactory.getLogger(LwM2mNodeSenMLJsonDecoder.class);
+
+    public static <T extends LwM2mNode> T decode(byte[] content, LwM2mPath path, LwM2mModel model, Class<T> nodeClass)
+            throws CodecException {
+        String jsonStrValue = content != null ? new String(content) : "";
+        SenMLPack pack = SenMLJson.fromJsonSenML(jsonStrValue);
+        return parseSenMLPack(pack, path, model, nodeClass);
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T extends LwM2mNode> T parseSenMLPack(SenMLPack pack, LwM2mPath path, LwM2mModel model,
+            Class<T> nodeClass) throws CodecException {
+
+        LOG.trace("Parsing SenML JSON object for path {}: {}", path, pack);
+
+        // Extract baseName
+        LwM2mPath baseName = extractAndValidateBaseName(pack, path);
+        if (baseName == null)
+            baseName = path; // if no base name, use request path as base name
```

Comment:
About "request path as basename, if basename is absent",  I read quickly the specification [§7.4.4 JSON](http://www.openmobilealliance.org/release/LightweightM2M/V1_1-20180710-A/HTML-Version/OMA-TS-LightweightM2M_Core-V1_1-20180710-A.html#7-4-4-0-744-JSON), I didn't find anything about that.

It seems this part has changed with the new JSON format. I guess this is to respect more the SenML rfc. (If this is the case, I approve this change)

Anyway, I maybe missed it so if you find anything let me know.

Reply:
Read-Composite is introduced in LwM2M 1.1. Table: 7.4.4.-6 JSON payload in the server request to read manufacturer name, battery level and registration lifetime shows an example of Read-Composite request like below: 

 [{"n":"/3/0/0"},
{"n":"/3/0/9"},
{"n":"/1/0/1"}]

In response to the above Read-Composite request the client will return a JSON payload as below:
[{"n":"/3/0/0", "vs":"Open Mobile Alliance"},
{"n" :"/3/0/9", "v":95},
{"n":"/1/0/1", "v":86400}]

In this case, the base name is absent, because of resources belong to different object instances, and object instances belong to different object. Do you think SenML JSON/CBOR encoder/decoder shall consider Read-Composite/Write-Composite/Observe-Composite now? Will appreciate if you share you opinion.

Reply:
I suspect our current API is not really compatible with this.
So I would go to let this aside for now :  one step at a time.

Reply:
(when I say "let this aside", I mean support of Read-Composite/Write-Composite/Observe-Composite)

Reply:
> (when I say "let this aside", I mean support of Read-Composite/Write-Composite/Observe-Composite)

Understood! Is there any plan or on-going contributing activity related with Composite operation now?

Reply:
There is no plan about that for now.
About on-going contribution, I just saw this https://github.com/eclipse/leshan/issues/563#issuecomment-466630438

Reply:
After extract base name from record array, I am trying to update value of name field of each of record to full path. Therefore, rest of logic can ignore base name anymore. Do you have any comments on that?

Reply:
I'm not sure to understand. I think we should remove `extractAndValidateBaseName(pack, path)` because we could face cases where there are several basenames ?

look at this https://tools.ietf.org/html/rfc8428#section-5.1.6
```
   [
     {"bn":"2001:db8::2/","bt":1.320078429e+09,
      "n":"temperature","u":"Cel","v":25.2},
     {"n":"humidity","u":"%RH","v":30},
     {"bn":"2001:db8::1/","n":"temperature","u":"Cel","v":12.3},
     {"n":"humidity","u":"%RH","v":67}
   ]
```

if we want a more LWM2M one we could imagine something like this : 
Reading object /2 (ACL) containing 2 instances
```
   [
     {"bn":"/2/0/","n":"0","v":3},
     {"n":"1","v":1},
     {"n":"3","v":123}
     {"bn":"/2/1/","n":"0","v":4},
     {"n":"1","v":1},
     {"n":"3","v":124}
   ]
```
(we could add unit tests like those)

So basename would be now handle in `groupRecordsByInstanceId`, maybe like this :
```java
    private static Map<Integer, Collection<SenMLRecord>> groupRecordsByInstanceId(Collection<SenMLRecord> records)
            throws CodecException {
        Map<Integer, Collection<SenMLRecord>> result = new HashMap<>();

        String basename = "";
        for (SenMLRecord record : records) {
            // update basename if present
            if (record.getBaseName() != null)
                basename = record.getBaseName();

            // Build resource path
            String path = record.getName() == null ? basename : basename + record.getName();
            LwM2mPath nodePath = new LwM2mPath(path);

            // Validate path
            if (!nodePath.isResourceInstance() && !nodePath.isResource()) {
                throw new CodecException(
                        "Invalid path [%s] for resource, it should be a resource or a resource instance path",
                        nodePath);
            }

            // Get SenML records for this instance
            Collection<SenMLRecord> recordForInstance = result.get(nodePath.getObjectInstanceId());
            if (recordForInstance == null) {
                recordForInstance = new ArrayList<>();
                result.put(nodePath.getObjectInstanceId(), recordForInstance);
            }

            // Add it to the list
            recordForInstance.add(record);
        }

        return result;
    }
```